proposal/package-importer.d.ts.md

```diff
@@ -0,0 +1,529 @@
+default Sass export for npm packages. For the most part, packages contain both
+JavaScript and styles, and use the `main` or `module` root keys to define the
+JavaScript entry point. Some packages use the `"sass"` key at the root of their
+`package.json`.
+
+Other packages have adopted [conditional exports], driven by build tools like
+[Vite], [Parcel] and [Sass Loader for Webpack] which all resolve Sass paths
+using the `"sass"` and the `"style"` custom conditions.
+
+[conditional exports]: https://nodejs.org/api/packages.html#conditional-exports
+[Vite]: https://github.com/vitejs/vite/pull/7817
+[Parcel]: https://github.com/parcel-bundler/parcel/blob/2d2400ded4615375ee6bd53ef77b4857ad1591dd/packages/transformers/sass/src/SassTransformer.js#L163
+[Sass Loader for Webpack]: https://github.com/webpack-contrib/sass-loader/blob/02df41203adfda96959e56abb43bd35a89ec11ba/src/utils.js#L514
+
+Because use of conditional exports is flexible and recommended for modern
+packages, this will be the primary method used for the Node package importer. We
+will support both the `"sass"` and the `"style"` conditions, as Sass can also
+use the CSS exports exposed through `"style"`. While in practice, `"style"`
+tends to be used solely for `css` files, we will support `scss`, `sass` and
+`css` files for either `"sass"` or `"style"`.
+
+While conditional exports allows package authors to define specific aliases to internal
+files, we will still use the Sass conventions for resolving file paths with
+partials, extensions and indices to discover the intended export alias. However,
+we will not apply that logic to the destination, and will expect library authors
+to map the export to the correct place. In other words, given a `package.json`
+with `exports` as below, The Node package importer will resolve a
+`@use "pkg:pkgName/variables";` to the destination of the `_variables.scss` export.
+
+```json
+{
+  "exports": {
+    "_variables.scss": {
+      "sass": "./src/sass/_variables.scss"
+    }
+  }
+}
+```
+
+## Types
+
+```ts
+import {Importer} from '../spec/js-api/importer';
+```
+
+### `nodePackageImporter`
+
+```ts
+export declare const nodePackageImporter: Importer;
+```
+
+#### The `nodePackageImporter` Importer
+
+The `nodePackageImporter` is a stubbed [Importer] that is replaced with an
+instantiated [Node Package Importer] at compile time.
+
+* If `nodePackageImporter.canonicalize` is invoked, throw an error.
+
+* If `nodePackageImporter.load` is invoked, throw an error.
+
+[Node Package Importer]: #node-package-importer
+
+#### Instantiating a `nodePackageImporter`
+
+After the first bullet points in [`compile`] and [`compileString`] in the
+Javascript Compile API, insert:
+
+* If any object in `options.importers` is exactly equal to the object
+  `nodePackageImporter`:
+
+  * Let `pkgImporter` be a [Node Package Importer] with an associated
+    `entryPointURL` of `require.main.filename`.
```

Comment:
When running node REPL `require.main` is `null`. Node has a hack internally to make the import work by adding the working directory as the local search path:

https://github.com/nodejs/node/blob/e7618fb5a5fc25d76b6474e2a6607f04fd6f10e0/lib/internal/modules/cjs/loader.js#L824-L832

I suppose we can set `entryPointURL` to `file://${process.cwd()}/` (node the slash at the end is important), which will give us enough information on the compiler side.

Reply:
@ntkme I updated this for the Embedded Protocol in https://github.com/sass/sass/pull/3660/commits/3e3c5f3d9fe30035c6ee1d3bc4e3b2e40c762880. Thanks!